File: gaussianTopHat.m
function imageFilt = gaussianTopHat(images,gaussSig)
% preprocess raw data to smooth out filamentous structures

% INPUT:
% images: an image sequence
% gaussSig: sigma (pixels) for gaussian smoothing with imgaussfilt
% OUTPUT:
% gaussian smoothed tophat filtered image sequences
% gaussSig=1.5;

    se = strel('disk',12);
    numFrames = size(images,3);
    imageFilt = zeros(size(images));
    for tt=1:numFrames
        tempFrame = images(:,:,tt);
        original = mat2gray(imgaussfilt(tempFrame,gaussSig));
        %in order to tophat filter, the image must be grayscale or binary
        tophatFiltered = imtophat(original,se);
        imageFilt(:,:,tt)=tophatFiltered;
    end

%
% figure()
% subplot(1,2,1)
% imshow(imadjust(im2uint8(tophatFiltered)));
% subplot(1,2,2)
% imshow(mat2gray(tempData(125:175,500:550,10)))
end

File: gaussianTopHat.py
import numpy as np
from skimage.morphology import disk
from skimage.filters import gaussian, rank
from skimage.util import img_as_ubyte

def gaussian_top_hat(images, gauss_sig):
    selem = disk(12)
    num_frames = images.shape[2]
    image_filt = np.zeros_like(images)

    for tt in range(num_frames):
        temp_frame = images[:, :, tt]
        original = img_as_ubyte(gaussian(temp_frame, sigma=gauss_sig))
        tophat_filtered = rank.tophat(original, selem=selem)
        image_filt[:, :, tt] = tophat_filtered

    return image_filt

# Example usage:
# Replace 'your_images' and 'your_gauss_sig' with actual values
your_images = np.random.rand(100, 100, 10)
your_gauss_sig = 1.5

result_image_filt = gaussian_top_hat(your_images, your_gauss_sig)

# Display the first frame of the resulting sequence
import matplotlib.pyplot as plt
plt.imshow(result_image_filt[:, :, 0], cmap='gray')
plt.title('First Frame - Gaussian Top Hat Filtered')
plt.show()
